Ok so this is sort of a big change :)

atlantis opened the pull request:
My etcd project requires TLS-authenticated gRPC connections, soooo I swapped out the http2 code with https://github.com/ysbaddaden/http2 (which seems like it will enter the stdlib at some point, at which point we could simply remove the http2 dependency and maybe tweak the client a bit if that's included?).

Configuration is probably the biggest change - switched to more like Kemal's syntax in order to allow runtime configuration/reconfiguration:

 ```
GRPC::Config.defaults do |config|
  ssl_context = OpenSSL::SSL::Context::Client.from_hash(
    {
      "verify_mode" => "none",
      "key"         => "cfssl/client1-key.pem",
      "cert"        => "cfssl/client1.pem",
      "ca"          => "cfssl/ca.pem",
    }
  )

  config.http2 = HTTP2::Client.new("127.0.0.1", 9090, ssl_context)
end

MaintenanceService = Etcdserverpb::Maintenance::Stub.new
pp MaintenanceService.status(::Etcdserverpb::StatusRequest.new)
```

In addition to global defaults, you can override the configuration for each service, if you wish. BTW I know this pull request is ridiculously huge and adds breaking changes, but it seems like using (soon-to-be) standard http2 code and adding TLS support and dynamic configuration will make this shard even more useful?

Thanks for a great library, and please let me know if you want anything changed and I'm happy to update the MR!

atlantis committed:
Seems to work (tested client, server, and generator)

atlantis commented:
Whoops hang on sorry - need to add HTTP2 trailer records into the mix (hopefully get the http2 shard to expose them).

jgaskins commented:
> My etcd project requires TLS-authenticated gRPC connections

I've been having the same issue wanting to build Grafana plugins. So, I hear you. :sob:

> I swapped out the http2 code with https://github.com/ysbaddaden/http2

THANK YOU! 💯 I noticed just last week that he'd started working on it again back in July and I was going to swap out the dependency here, too. I didn't _want_ to maintain an HTTP2 implementation, but for various reasons that one didn't work as a shard at the time that I wrote this.

> Configuration is probably the biggest change - switched to more like Kemal's syntax in order to allow runtime configuration/reconfiguration
> …
> In addition to global defaults, you can override the configuration for each service, if you wish.

This is interesting. I designed this shard around Google's gRPC clients in languages like Ruby and JavaScript where you specify what host to connect to on a per-stub basis.

This implementation is more similar to how the Go client maintains a single gRPC client and passes that client to stub constructors, which may be more in-line with how gRPC (and HTTP/2 more broadly) should work in practice for real-world apps.

> I know this pull request is ridiculously huge and adds breaking changes, but it seems like using (soon-to-be) standard http2 code and adding TLS support and dynamic configuration will make this shard even more useful?

I haven't had a chance to look much into the code yet but, honestly, I do like that this makes things more flexible and gets the HTTP/2 implementation off of my plate. You only added 221 lines despite swapping the entire backend out. That's pretty reasonable, imo. 😄

I'd love to ease the transition from the old implementation to this one. If there's a way we could add some sort of compatibility layer, even if it's short-lived and immediately marked as [`@[Deprecated]`](https://crystal-lang.org/api/1.18.2/Deprecated.html) (wow, the docs page for that annotation looks really bad), I think that would be ideal. If that's not feasible, I can pretend I do not see it.

> Whoops hang on sorry - need to add HTTP2 trailer records into the mix (hopefully get the http2 shard to expose them).

lol Yes, I had the same problem, so I opened [this PR](https://github.com/ysbaddaden/http2/pull/15). It's outdated since the recent changes to the shard, but yes, definitely a thing we need!

atlantis commented:
Ah ok thanks... for posterity, I did:

```
http2.send(headers, io.to_slice) do |stream|
  response = stream.data
  if raw_status = (stream.headers["grpc-status"]? || stream.trailers["grpc-status"]?).try(&.to_i?)
    status_code = GRPC::StatusCode.from_value?(raw_status) || GRPC::StatusCode::UNKNOWN
    if status_code.ok?
      compressed = response.read_byte != 0 # TODO: Handle compression?
      length = response.read_bytes Int32, IO::ByteFormat::BigEndian

      return \{{response_type}}.from_protobuf(response)
    else
      raise GRPC::BadStatus.new(
        status_code,
        stream.headers["grpc-message"]? || stream.trailers["grpc-message"]? || "Unknown error!"
      )
    end

  else
    raise "Unable to get status code from response"
  end
end
```
... and it seemed odd that `headers` and `trailers` didn't operate the same way. Thanks for the response!

atlantis committed:
smooth transition by matching old config pattern and check trailers for grpc-status

atlantis commented:
Thanks for the quick response! Yeah the way etcd does things there are like 6 different gRPC services you have to interact with (and in common usage you'll interact with at least 2-3 of them at a time), so having some way to share an HTTP2 connection seems ideal, since it's particularly good at sharing requests over a single connection!

Added a compatibility layer as suggested, and on that note: I originally tried to leave the API for that alone and just add tls_context to the Service::Stub initializer, and what happened was pretty funny:

```
ssl_context = OpenSSL::SSL::Context::Client.from_hash(
    {
      "verify_mode" => "none",
      "key"         => "cfssl/client1-key.pem",
      "cert"        => "cfssl/client1.pem",
      "ca"          => "cfssl/ca.pem",
    }
  )

EtcdService = Etcdserverpb::Maintenance::Stub.new("127.0.0.1", 2379, ssl_context)
```
=> `Error: undefined local variable or method 'ssl_context' for top-level`

I suppose this's because those params need to be available at macro compilation time rather than execution time? So anyway there's a bit of context on why this MR ended up changing the config pattern 😆

atlantis committed:
hmm in the real world we'll need the ability to set socket timeouts

atlantis committed:
When including in another project seems to need the namespacing for some reason

atlantis committed:
Don't pollute HTTP2 namespace and allow default_headers

Use GRPC::Client instead of HTTP2::Client (someday that will be a think hopefully and we can leverage it).

Also allow GRPC::Client to have default headers to handle cases like token-based auth, and make sure all the headers sort properly (pseudo-headers starting with `:` have to be at the very top according to the spec).

atlantis committed:
Fix client concurrency

Need to test streaming but unary requests work consistently now

atlantis committed:
Add streaming to http2 client

atlantis committed:
Add helpers to grpc encode/decode protobufs

atlantis committed:
Do the safe thing until http2::client provides a legit streaming api

james-hill commented:
Ok in case it's helpful there's now a working example of how this could be used at: https://github.com/place-labs/crystal-etcd/pull/27 to switch from the etcd REST API to the GRPC API.

atlantis committed:
Don't include the http2/server stuff cause it breaks kemal

atlantis committed:
Use fixed http2 shard for now

atlantis committed:
Close streams propertly

atlantis committed:
Add unit tests and pull new http2 version without memory leaks

atlantis committed:
Properly close fibers on connection drop
